runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/FanOutStreamingEngineWorkerHarness.java

```diff
@@ -256,58 +242,132 @@ private GetDataStream getGlobalDataStream(String globalDataKey) {
-                closeStaleStreamsAndCreateNewStreams(newWindmillConnections.values()))
+            .setWindmillStreams(newStreams)
             .setGlobalDataStreams(
                 createNewGlobalDataStreams(newWindmillEndpoints.globalDataEndpoints()))
             .build();
-
     LOG.info(
         "Setting new connections: {}. Previous connections: {}.",
         newConnectionsState,
         connections.get());
     connections.set(newConnectionsState);
-    isBudgetRefreshPaused.set(false);
-    getWorkBudgetRefresher.requestBudgetRefresh();
+    getWorkBudgetDistributor.distributeBudget(newStreams.values(), totalGetWorkBudget);
+    activeMetadataVersion = newWindmillEndpoints.version();
+    LOG.info(
+        "Consumed new endpoints. previous metadata version: {}, current metadata version: {}",
+        previousMetadataVersion,
+        activeMetadataVersion);
+  }
+
+  /** Close the streams that are no longer valid asynchronously. */
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void closeStaleStreams(
+      Collection<WindmillConnection> newWindmillConnections,
+      ImmutableMap<WindmillConnection, WindmillStreamSender> currentStreams) {
+    currentStreams.entrySet().stream()
+        .filter(
+            connectionAndStream -> !newWindmillConnections.contains(connectionAndStream.getKey()))
+        .forEach(
+            entry ->
+                CompletableFuture.runAsync(
+                    () -> {
+                      LOG.debug("Closing streams to {}", entry);
+                      try {
+                        entry.getValue().closeAllStreams();
+                        entry
+                            .getKey()
+                            .directEndpoint()
+                            .ifPresent(channelCachingStubFactory::remove);
+                        LOG.debug("Successfully closed streams to {}", entry);
+                      } catch (Exception e) {
+                        LOG.error("Error closing streams to {}", entry);
+                      }
+                    },
+                    windmillStreamManager));
+  }
+
+  private synchronized CompletableFuture<ImmutableMap<WindmillConnection, WindmillStreamSender>>
+      createAndStartNewStreams(Collection<WindmillConnection> newWindmillConnections) {
+    ImmutableMap<WindmillConnection, WindmillStreamSender> currentStreams =
+        connections.get().windmillStreams();
+    return MoreFutures.allAsList(
+            newWindmillConnections.stream()
+                .map(
+                    connection ->
+                        MoreFutures.supplyAsync(
+                            () ->
+                                Pair.of(
+                                    connection,
+                                    Optional.ofNullable(currentStreams.get(connection))
+                                        .orElseGet(
+                                            () -> createAndStartWindmillStreamSender(connection))),
```

Comment:
startStream() does block.  It will block until the header is successfully sent. 
https://github.com/apache/beam/blob/d4dd58b2c4c4b5867ace4bdd34e1bcc32de963cc/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/AbstractWindmillStream.java#L182

onNewStream() blocks until a header is successfully sent and will keep retrying until it is or the stream is closed/shutdown